Spec: Document Snapshot Summary Optional Fields for Standardization (#11660)
apache/iceberg: format/spec.md

The snapshot's `first-row-id` is the starting `first_row_id` assigned to manifests in the snapshot's manifest list.

#### Snapshot Summary

##### Required Field `operation`
The snapshot summary's `operation` field is used by some operations, like snapshot expiration, to skip processing certain snapshots. Possible `operation` values are:

*   `append` -- Only data files were added and no files were removed.
*   `replace` -- Data and delete files were added and removed without changing table data; i.e., compaction, changing the data file format, or relocating data files.
*   `overwrite` -- Data and delete files were added and removed in a logical overwrite operation.
*   `delete` -- Data files were removed and their contents logically deleted and/or delete files were added to delete rows.

> ##### Optional Metrics

Reviewer: As on the proposal doc I think this would be a nice appendix table

> All metrics fields should have numeric string values (e.g., `"120"`).

Reviewer: I think we should note that these values may be used by engines for optimizations so they must be correct but they can be skipped if an engine doesn't want to write them. Maybe that's clear enough already though with "optional"

Some of them are also used to represent partition-level metrics, in [Optional Partition-Level Summary](#optional-partition-level-summary).

| Field                               | Description                                                       | Used in Partition-Level Summary |

> |-------------------------------------|-------------------------------------------------------------------|---------------------------------|

Reviewer: I'm not sure "current" is required here

| **`added-data-files`**              | Number of data files added in the current snapshot                | Yes                             |
| **`deleted-data-files`**            | Number of data files deleted in the current snapshot              | Yes                             |

> | **`total-data-files`**              | Total number of data files in the current snapshot                | No                              |

Reviewer: Total number of live data files?

> | **`added-delete-files`**            | Number of delete files added in the current snapshot              | Yes                             |

Reviewer: Is this a combo of position and equality deletes

| **`added-equality-delete-files`**   | Number of equality delete files added in the current snapshot     | Yes                             |
| **`removed-equality-delete-files`** | Number of equality delete files removed in the current snapshot   | Yes                             |
| **`added-position-delete-files`**   | Number of position delete files added in the current snapshot     | Yes                             |
| **`removed-position-delete-files`** | Number of position delete files removed in the current snapshot   | Yes                             |
| **`added-dvs`**                     | Number of deletion vectors added in the current snapshot          | Yes                             |
| **`removed-dvs`**                   | Number of deletion vectors removed in the current snapshot        | Yes                             |
| **`removed-delete-files`**          | Number of delete files removed in the current snapshot            | Yes                             |

> | **`total-delete-files`**            | Total number of delete files in the current snapshot              | No                              |

Reviewer: This is a combination of dv, df and eq deletes right?

| **`removed-files-size`**            | The size of files removed in the current snapshot                 | Yes                             |
| **`total-files-size`**              | The size of all files in the current snapshot                     | No                              |
| **`added-position-deletes`**        | Number of position delete records added in the current snapshot   | Yes                             |
| **`removed-position-deletes`**      | Number of position delete records removed in the current snapshot | Yes                             |
| **`total-position-deletes`**        | Total number of position delete records in the current snapshot   | No                              |
| **`added-equality-deletes`**        | Number of equality delete records added in the current snapshot   | Yes                             |
| **`removed-equality-deletes`**      | Number of equality delete records removed in the current snapshot | Yes                             |
| **`total-equality-deletes`**        | Total number of equality delete records in the current snapshot   | No                              |

> | **`deleted-duplicate-files`**       | Number of duplicate files deleted in the current snapshot         | No                              |

Reviewer: Unclear where duplicate files are coming from

> | **`changed-partition-count`**       | Number of partitions changed in the current snapshot              | No                              |

Reviewer: I think we need to detail, partitions with added, deleted or modified files?

| **`added-position-deletes`**        | Number of position delete records added in the snapshot                                          |
| **`removed-position-deletes`**      | Number of position delete records removed in the snapshot                                        |
| **`total-position-deletes`**        | Total number of position delete records in the snapshot                                          |
| **`added-equality-deletes`**        | Number of equality delete records added in the snapshot                                          |
| **`removed-equality-deletes`**      | Number of equality delete records removed in the snapshot                                        |
| **`total-equality-deletes`**        | Total number of equality delete records in the snapshot                                          |
| **`deleted-duplicate-files`**       | Number of duplicate files deleted (duplicates are files recorded more than once in the manifest) |
| **`changed-partition-count`**       | Number of partitions with files added or removed in the snapshot                                 |

> #### Other Fields

Reviewer 2: There's also `app-id` from `SparkCatalog`.

Author: I think we can skip it since `app-id` is an engine/service specific detail. https://github.com/apache/iceberg/pull/11660#discussion_r1919172817

Reviewer 2: Okay, I missed that discussion.

When processing point in time queries implementations should use "snapshot-log" metadata to lookup the table state at the given point in time. This ensures time-travel queries reflect the state of the table at the provided timestamp. For example a SQL query like `SELECT * FROM prod.db.table TIMESTAMP AS OF '1986-10-26 01:21:00Z';` would find the snapshot of the Iceberg table just prior to '1986-10-26 01:21:00 UTC' in the snapshot logs and use the metadata from that snapshot to perform the scan of the table. If no  snapshot exists prior to the timestamp given or "snapshot-log" is not populated (it is an optional field), then systems should raise an informative error message about the missing metadata.

> ## Appendix G: Optional Snapshot Summary Fields

Reviewer 3: Why an appendix rather than adding this to the snapshot summary content? This states that "Metrics must be accurate if written" so it adds requirements that probably aren't appropriate for an appendix.

Author: Thanks for all the suggestions! I removed the statement and also took Ryan's suggestion to make these fields a subsection of Appendix F - Implementation details, which states clear that it should be for recommendation/reference instead of requirements.

Snapshot summary can include metrics fields to track numeric stats of the snapshot (see [Metrics](#metrics)) and operational details (see [Other Fields](#other-fields)). The value of these fields should be of string type (e.g., `"120"`).

### Metrics

> Metrics must be accurate if written, as engines may rely on them for optimization.

Reviewer 3: Style: there should be an empty newline following header lines.

| **`added-position-deletes`**        | Number of position delete records added in the snapshot                                          |
| **`removed-position-deletes`**      | Number of position delete records removed in the snapshot                                        |
| **`total-position-deletes`**        | Total number of position delete records in the snapshot                                          |
| **`added-equality-deletes`**        | Number of equality delete records added in the snapshot                                          |
| **`removed-equality-deletes`**      | Number of equality delete records removed in the snapshot                                        |
| **`total-equality-deletes`**        | Total number of equality delete records in the snapshot                                          |
| **`deleted-duplicate-files`**       | Number of duplicate files deleted (duplicates are files recorded more than once in the manifest) |
| **`changed-partition-count`**       | Number of partitions with files added or removed in the snapshot                                 |

> ### Other Fields

Reviewer 2: Other fields also include `engine-name`, `engine-version`, `app-id`.

Author: I was on the edge of including these since they are not constants in `SnapshotSummary.java`. `engine-name` and `engine-version` seem fine as they’re unambiguous and included in the view's summary spec: https://iceberg.apache.org/view-spec/#summary. However, `app-id` feels too generic and poorly defined, so I’ve left it out.

I've added rows for `engine-name` and `engine-version`. Please let me know what you think.

Reviewer 4: Spark adds the `spark.app.id` to the snapshot summary as seen here 
https://iceberg.apache.org/docs/1.6.0/spark-queries/#snapshots

Reviewer: IMHO, We should skip anything that's engine specific since the goal is to provide definitions for things that are universal and may be used by different engines. So we don't need spark.app.id here because no other engine needs a consistent definition of that.

| **`added-equality-deletes`**        | Number of equality delete records added in the snapshot                                          |
| **`removed-equality-deletes`**      | Number of equality delete records removed in the snapshot                                        |
| **`total-equality-deletes`**        | Total number of equality delete records in the snapshot                                          |
| **`deleted-duplicate-files`**       | Number of duplicate files deleted (duplicates are files recorded more than once in the manifest) |
| **`changed-partition-count`**       | Number of partitions with files added or removed in the snapshot                                 |

#### Other Fields

| Field                    | Example    | Description                                                     |
|--------------------------|------------|-----------------------------------------------------------------|
| **`wap.id`**             | "12345678" | The Write-Audit-Publish id of a staged snapshot                 |
| **`published-wap-id`**   | "12345678" | The Write-Audit-Publish id of a snapshot already been published |

> | **`source-snapshot-id`** | "12345678" | The id of the snapshot picked to be cherry-picked               |

Reviewer: Probably work on this description?
```suggestion
| **`source-snapshot-id`** | "12345678" | The original id of a cherry-picked snapshot            |              
```

> | **`replace-partitions`** | "true"     | Whether the operation is a `ReplacePartitions`[1]               |

Reviewer: Is this universal enough of a concept? I may skip this one just because it feels like it should have been in "operation" if it was really important

Author: linking my reply above: https://github.com/apache/iceberg/pull/11660#discussion_r1902124366

Let's just skip this one

When processing point in time queries implementations should use "snapshot-log" metadata to lookup the table state at the given point in time. This ensures time-travel queries reflect the state of the table at the provided timestamp. For example a SQL query like `SELECT * FROM prod.db.table TIMESTAMP AS OF '1986-10-26 01:21:00Z';` would find the snapshot of the Iceberg table just prior to '1986-10-26 01:21:00 UTC' in the snapshot logs and use the metadata from that snapshot to perform the scan of the table. If no  snapshot exists prior to the timestamp given or "snapshot-log" is not populated (it is an optional field), then systems should raise an informative error message about the missing metadata.

> ## Appendix G: Optional Snapshot Summary Fields
> 
> ### Metrics
> Snapshot summary can include metrics fields to track numeric stats of the snapshot. The value of these fields should be numeric strings (e.g., `"120"`).
> Some of them are also used to represent partition-level metrics, in [Partition-Level Summary](#partition-level-summary).
> Metrics must be accurate if written, as engines may rely on them for optimization.

Reviewer 4: nit: wydt about moving this description up one level, between "Appendix G" and "Metrics" since this is describing generally what fields belong to the Snapshot Summary. It should apply to all subheaders ( "Metrics", "Partition-Level", "Other")

```suggestion
## Appendix G: Optional Snapshot Summary Fields
Snapshot summary can include metrics fields to track numeric stats of the snapshot, see [Metrics](#metrics). 
Some of them are also used to represent partition-level metrics, in [Partition-Level Summary](#partition-level-summary).

The value of these fields should be of string type (e.g., `"120"`).

### Metrics
Metrics must be accurate if written, as engines may rely on them for optimization.
```

Author: Thanks for the suggestion! I've moved the the description up one level and make it mention all the sub-sections. I removed the "partition-level summary" based on Russell's suggestion above.

| **`total-delete-files`**            | Total number of live positional/equality delete files and deletion vectors in the snapshot            | No                              |
| **`added-records`**                 | Number of records added in the snapshot                                                               | Yes                             |
| **`deleted-records`**               | Number of records deleted in the snapshot                                                             | Yes                             |
| **`total-records`**                 | Total number of records in the snapshot                                                               | No                              |
| **`added-files-size`**              | The size of files added in the snapshot                                                               | Yes                             |
| **`removed-files-size`**            | The size of files removed in the snapshot                                                             | Yes                             |

> | **`total-files-size`**              | The size of all files in the snapshot                                                                 | No                              |

Reviewer 4: ```suggestion
| **`total-files-size`**              | Total size of files added in the snapshot                                                                 | No                              |
```
align with other `total-` descriptions

| **`added-position-deletes`**        | Number of position delete records added in the snapshot                                               | Yes                             |
| **`removed-position-deletes`**      | Number of position delete records removed in the snapshot                                             | Yes                             |
| **`total-position-deletes`**        | Total number of position delete records in the snapshot                                               | No                              |
| **`added-equality-deletes`**        | Number of equality delete records added in the snapshot                                               | Yes                             |
| **`removed-equality-deletes`**      | Number of equality delete records removed in the snapshot                                             | Yes                             |
| **`total-equality-deletes`**        | Total number of equality delete records in the snapshot                                               | No                              |

> | **`deleted-duplicate-files`**       | Number of duplicate files deleted, where duplicates are files recorded more than once in the manifest | No                              |

Reviewer 4: ```suggestion
| **`deleted-duplicate-files`**       | Number of duplicate files deleted (duplicates are files recorded more than once in the manifest) | No                              |
```

| **`changed-partition-count`**       | Number of partitions with files added or removed in the snapshot                                      | No                              |

> ### Partition-Level Summary

Reviewer: I'm not sure we should document this, the string representation we use for the path is not really well documented and in general we probably don't want folks including partition level summaries. Additionally it references a Table property (write.summary.partition-limit) which I really would like to avoid. I think we should stick to the table level stats above and remove this section and the associated column.

Author: Good point! I also found it challenging to document partition-level summaries while writing the appendix, as they require explanations for a lot of context. Referencing `write.summary.partition-limit` directly in the table spec also feels out of place. Additionally, most users won’t encounter partition-level summaries because the default value for `write.summary.partition-limit` is `0`.

Even so, I think it’s still helpful to document this information somewhere for those who might enable or work with partition-level summaries. One option could be to move the description to the [Table Configuration](https://iceberg.apache.org/docs/nightly/configuration/#write-properties) section and include a note under write.summary.partition-limit to explain what partition-level summaries mean. This approach ties the information to the property users would need to modify if they choose to enable this feature. The downside is that snapshot summaries would end up being documented in two places, but this might be a reasonable tradeoff to ensure clarity and usability.

Do you think this sound like a better idea?

Reviewer: I think we should just skip anything that is related directly to a table property? I think I am ok with just not adding those ones

*   `partitions.<partition path>`: A prefix for partition-level metrics, appended with the string representation of the partition path[1]. The values are string maps containing partition metrics, which include some fields from [Metrics](#metrics).

Notes:

1. Partition paths are written as Hive partition paths in the format `<partition field name>=<partition value>`. For multiple partitions, they are separated by `/` (e.g. `year=2023/month=12/day=19`).

### Other Fields

| Field                    | Example    | Description                                                     |
|--------------------------|------------|-----------------------------------------------------------------|
| **`wap.id`**             | "12345678" | The Write-Audit-Publish id of a staged snapshot                 |
| **`published-wap-id`**   | "12345678" | The Write-Audit-Publish id of a snapshot already been published |
| **`source-snapshot-id`** | "12345678" | The id of the snapshot picked to be cherry-picked               |

> | **`replace-partitions`** | `true`     | Whether the operation is a `ReplacePartitions`                  |

Author: Thanks for catching this! Let me add more context for `ReplacePartitions` here. 

That said, I’m on the fence about including this field. Most users likely won’t encounter it, as `ReplacePartitions` is generally not recommended. However, I decided to include it to make the appendix more comprehensive and to "reserve" the field name to prevent potential conflicts.

Link for reference: [ReplacePartitions.java](https://github.com/apache/iceberg/blob/c07f2aabc0a1d02f068ecf1514d2479c0fbdd3b0/api/src/main/java/org/apache/iceberg/ReplacePartitions.java#L22-L38)
